samples/echo/src/main/cpp/AudioEffect.cpp

```diff
@@ -0,0 +1,277 @@
+    bgMixFactorInt_ = (int32_t)
+      (bgMixFactor_ * kFloatToIntMapFactor + 0.5f);
+    fgMixFactorInt_ = kFloatToIntMapFactor - bgMixFactorInt_;
+  }
+}
+
+/**
+ * Insert a raw PCM audio buffer to blend with live audio
+ *
+ * @param samples points to PCM audio buffer
+ * @param sampleCount is total samples pointed by samples
+ * @param channelCount channels for PCM audio pointed by samples
+ * @param freq is PCM audio frequency (48000hz for this sample)
+ */
+void  AudioMixer::addStream(std::unique_ptr<int16_t[]>samples, size_t sampleCount,
+          int32_t sampleRate, int32_t channelCount, oboe::AudioFormat format){
+  if (busy_) {
+    LOGW("filtering in progress, filter configuration is IGNORED");
+    return;
+  }
+  bgAudio_ = std::move(samples);
+  bgAudioSampleCount_ = sampleCount;
+  sampleRate_ = sampleRate;
+  format_ = format;
+  channelCount_ = channelCount;
+
+  curPosition_ = 0;
+}
+
+/**
+ * Adding audio processing into the live audio
+ * @param liveAudio is recorded audio
+ * @param samplesPerFrame is same as channelCount.
+ * @param numFrames represents frames pointed by liveAudio
+ *        total samples = numFrames * samplesPerFrame
+ */
+void AudioMixer::process(int16_t *liveAudio, int32_t channelCount,
+                          int32_t numFrames) {
+  assert(bgAudio_ && liveAudio);
+  if (numFrames > bgAudioSampleCount_ || channelCount != channelCount_ ||
+      bgMixFactorInt_ == 0) {
+    return;
+  }
+
+  busy_ = true;
+  int32_t curSample;
+  for (int i = 0; i < (numFrames * channelCount); i++) {
+    curSample = liveAudio[i];
+    curSample = curSample * fgMixFactorInt_ +
+                bgAudio_[curPosition_] * bgMixFactorInt_;
+    curSample /= kFloatToIntMapFactor;
+
+    curSample = (curSample > SHRT_MAX ? SHRT_MAX : curSample);
+    liveAudio[i] = (int16_t)(curSample < SHRT_MIN ? SHRT_MIN : curSample);
+    curPosition_ = (curPosition_ + 1 ) % bgAudioSampleCount_;
+  }
+  busy_ = false;
+}
+
+/**
+ * query for audio format supportability
+ */
+bool AudioMixer::AudioFormatSupported(int32_t frequency,
+                  int32_t channels, oboe::AudioFormat format) const {
+  return (frequency == sampleRate_ &&
+          channels == channelCount_ &&
+          format == format_);
+}
+
+/**
+ * Constructor for AudioDelay
+ * @param sampleRate
+ * @param channelCount
+ * @param format
+ * @param delay
+ * @param decay
+ */
+AudioDelay::AudioDelay(int32_t sampleRate,
+                       int32_t channelCount,
+                       oboe::AudioFormat format,
+                       float  delay,
+                       float  decay) :
+    AudioFormat(sampleRate, channelCount, format),
+    delay_(delay), decay_(decay) {
+
+  feedbackFactor_ = static_cast<int32_t>(decay_ * kFloatToIntMapFactor);
+  liveAudioFactor_ = kFloatToIntMapFactor - feedbackFactor_;
+  allocateBuffer();
+}
+
+/**
+ * Destructor
+ */
+AudioDelay::~AudioDelay() {
+  if(buffer_) delete static_cast<uint8_t*>(buffer_);
```

Comment:
Two notes:

- `delete` actually checks whether the pointer is `nullptr`, so no need to do a check here;

- if the type of buffer is actually known (uint8_t), why store it as `void*`? You could instead store it as `std::unique_ptr<uint8_t[]>` and avoid manual deletion.

Reply:
the main reason of void* is, someday, have a generic filter class as a base class for different type of format. void* would be for that. It is up to the derived class to cast it back to its own thing.  This code is far from that, I will delete it.